src/llfs/slot_writer.hpp

```diff
@@ -225,6 +267,110 @@ class TypedSlotWriter<PackedVariant<Ts...>> : public SlotWriter
     }
   };
 
+  /** \brief An append operation of one or more slots.
+   */
+  class MultiAppend
+  {
+   public:
+    /** \brief Initialize a MultiAppend operation using the passed TypedSlotWriter.
+     *
+     * This will obtain a lock on the slot writer's LogDevice::Writer mutex.
+     */
+    explicit MultiAppend(TypedSlotWriter& slot_writer) noexcept : writer_lock_{slot_writer}
+    {
+    }
+
+    MultiAppend(const MultiAppend&) = delete;
+    MultiAppend& operator=(const MultiAppend&) = delete;
+
+    //----- --- -- -  -  -   -
+
+    Status begin_atomic_range(const batt::Grant& caller_grant) noexcept
+    {
+      return this->writer_lock_.begin_atomic_range(caller_grant);
+    }
+
+    Status end_atomic_range(const batt::Grant& caller_grant) noexcept
+    {
+      return this->writer_lock_.end_atomic_range(caller_grant);
+    }
+
+    /** \brief Packs a single slot into the log device buffer.  Does not commit the slot; all slots
+     * are committed atomically when this->finalize() is called.
+     */
+    template <typename T, typename PackedT = PackedTypeFor<T>>
+    StatusOr<SlotParseWithPayload<const PackedT*>> typed_append(const batt::Grant& caller_grant,
```

Comment:
I know templates are annoying to write out to .cpp files. Is it worth doing so in this case? This function is long for a .hpp file.

Reply:
Generally, it only works to put the definition of a template in a .cpp file if one knows ahead of time what all the instantiations of that template will be.  The compiler needs access to the full template definition for any types it hasn't seen yet at the point of instantiation.  In this case, it does need to be in some sort of exported header file because `T` and `PackedT` can be supplied by a downstream library.